### PR TITLE
FCBHDBP-515 enable non-integer verse identifier

### PR DESCRIPTION
**Pull request by vichugofsl**
# Description
The verse_start column has been changed from an integer type to a string type because sometimes a verse is represented as a string, such as '3a'. As a result, we have updated the verse_start column for the following entities: `bible_verses`, `bible_files`, `bible_file_timestamps`, `user_bookmarks`, `user_highlights`, and `user_notes`. Then, we have created a new column called `verse_sequence` of type integer for each above entity. As a result, we have replaced the column `verse_start` with `verse_sequence` in order to sort the records.

# NOTE 1
We need to run the following SQL statements to create the new column for all entities.

```sql
-- bible_verses
ALTER TABLE bible_verses add column verse_sequence tinyint unsigned NOT NULL;

UPDATE bible_verses SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE bible_verses DROP KEY unique_text_reference;

ALTER TABLE bible_verses modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL;

ALTER TABLE bible_verses ADD UNIQUE KEY `unique_text_reference` (`hash_id`,`book_id`,`chapter`,`verse_start`);
------------------------------------------------------------------------------------------------------------------------------------------------

-- bible_file_timestamps
ALTER TABLE bible_file_timestamps add column verse_sequence tinyint unsigned DEFAULT NULL;

UPDATE bible_file_timestamps SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE bible_file_timestamps modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL;
------------------------------------------------------------------------------------------------------------------------------------------------

-- bible_files
ALTER TABLE bible_files add column verse_sequence tinyint unsigned DEFAULT NULL;

UPDATE bible_files SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE bible_files modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL;
------------------------------------------------------------------------------------------------------------------------------------------------

-- user_notes
ALTER TABLE dbp_users.user_notes add column verse_sequence tinyint unsigned NOT NULL;

UPDATE dbp_users.user_notes SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE dbp_users.user_notes modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL;
------------------------------------------------------------------------------------------------------------------------------------------------

-- user_bookmarks
ALTER TABLE dbp_users.user_bookmarks add column verse_sequence tinyint unsigned NOT NULL;

UPDATE dbp_users.user_bookmarks SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE dbp_users.user_bookmarks modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL;
------------------------------------------------------------------------------------------------------------------------------------------------

-- user_highlights
ALTER TABLE dbp_users.user_highlights add column verse_sequence tinyint unsigned NOT NULL;

UPDATE dbp_users.user_highlights SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE dbp_users.user_highlights modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL;
------------------------------------------------------------------------------------------------------------------------------------------------

```

## Issue Link
https://fullstacklabs.atlassian.net/browse/FCBHDBP-515

## How Do I QA This

We should run the following postman tests and they should pass:
- [M] bible.is mobile requests
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/collection/12519377-87d1d0ee-d642-4fb7-ba80-606638d4e5e3?action=share&creator=17122915

- [M] Download requests
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/collection/12519377-3a68ded7-3a3d-40a3-9feb-93ce491311a8?action=share&creator=17122915

- Annotations tests
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/folder/12519377-3cc08c6e-6640-45c3-bc82-96d2d51162d3

- Notes
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-6ebd415f-3060-422b-8b9a-dcd5ee675028

### CODE REVIEW

File: app/Http/Controllers/User/BookmarksController.php
Comment (bradflood) [inline app/Http/Controllers/User/BookmarksController.php:168]: is this trying to cast a string to an int? If yes, then what will happen if verse_start is "3a"?

File: app/Http/Controllers/User/BookmarksController.php
Comment (bradflood) [inline app/Http/Controllers/User/BookmarksController.php:229]: same concern as above - casting string to int when string is "3a"

File: app/Http/Controllers/User/HighlightsController.php
Comment (bradflood) [inline app/Http/Controllers/User/HighlightsController.php:289]: see question above

**Review by bradflood — COMMENTED** [review]
SonarCloud failure is due to duplicate code, which only applies to the migration scripts.  Therefore, this will not block the PR.
Victor, I do have a question about PHP syntax... whether there is an attempt to cast a string to an int when the string might be for example "3a". If you can address that satisfactorily, this can be approved
I will run scripts manually after the PR is approved and before it is merged.

**Comment by vichugofsl** [comment]
> SonarCloud failure is due to duplicate code, which only applies to the migration scripts. Therefore, this will not block the PR. Victor, I do have a question about PHP syntax... whether there is an attempt to cast a string to an int when the string might be for example "3a". If you can address that satisfactorily, this can be approved I will run scripts manually after the PR is approved and before it is merged.

@bradflood about the PHP syntax, when a string is cast to an integer, any non-numeric characters will be removed, leaving only the numerical value.

E.g. "3a" => 3

I have created a postman test to validate it:
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-818d00db-875e-4cb0-ae84-68669d50c0a6

File: app/Http/Controllers/User/BookmarksController.php
Comment (vichugofsl) [reply app/Http/Controllers/User/BookmarksController.php:229]: @bradflood  when a string is cast to an integer, any non-numeric characters will be removed, leaving only the numerical value.

E.g. "2aedA" => 2

**Review by bradflood — APPROVED** [review]

**Pull request merged**
